appveyor.yml

```diff
@@ -20,6 +20,11 @@ environment:
       API_POLLER: poll
       WITH_LIBSODIUM: ON
       ENABLE_CURVE: ON
+    - platform: Win32
+      configuration: Release
+      POLLER: epoll
```

Comment:
could this be done in one of the existing builds? appveyor is already so slow...

Reply:
I still think we actually should have more jobs than we currently have. There might be 64-bit specific issues in the epoll code, e.g., which we wouldn't catch with the current configuration. Maybe we could merge jobs, such that a partial build output is reused: build with select poller first, then switch to epoll in the same job. However, if there is a failure then, it would be much harder to see where it occurred.

I suspected earlier that the cmake run is taking so long, but I only recently found out that timestamps are visible when you download the appveyor log, and I found that my assumption is true. As you can see here, the cmake calls takes almost 2 minutes:
```
[00:00:26] cmake -D CMAKE_INCLUDE_PATH="%SODIUM_INCLUDE_DIR%" -D CMAKE_LIBRARY_PATH="%SODIUM_LIBRARY_DIR%" -D WITH_LIBSODIUM="%WITH_LIBSODIUM%" -D ENABLE_DRAFTS="ON" -D ENABLE_CURVE="%ENABLE_CURVE%" -D API_POLLER="%API_POLLER%" -D POLLER="%POLLER%" -D CMAKE_C_FLAGS_RELEASE="/MT" -D CMAKE_C_FLAGS_DEBUG="/MTd" -D WITH_LIBSODIUM="%WITH_LIBSODIUM%" -D LIBZMQ_WERROR="ON" -G "%CMAKE_GENERATOR%" "%APPVEYOR_BUILD_FOLDER%"
[00:00:31] -- The C compiler identification is MSVC 18.0.40629.0
[00:00:33] -- The CXX compiler identification is MSVC 18.0.40629.0
[00:00:33] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe
[00:00:36] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe -- works
[00:00:36] -- Detecting C compiler ABI info
[00:00:38] -- Detecting C compiler ABI info - done
[00:00:38] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe
[00:00:40] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe -- works
[00:00:40] -- Detecting CXX compiler ABI info
[00:00:42] -- Detecting CXX compiler ABI info - done
[00:00:42] -- Detecting CXX compile features
[00:00:44] -- Detecting CXX compile features - done
[00:00:45] -- Performing Test COMPILER_SUPPORTS_CXX11
[00:00:47] -- Performing Test COMPILER_SUPPORTS_CXX11 - Failed
[00:00:47] -- Performing Test COMPILER_SUPPORTS_C11
[00:00:49] -- Performing Test COMPILER_SUPPORTS_C11 - Failed
[00:00:49] -- Found SODIUM: C:/projects/libsodium/bin/Win32/Release/v120/dynamic/libsodium.lib  
[00:00:49] -- Using libsodium for CURVE security
[00:00:49] -- Looking for kqueue
[00:00:51] -- Looking for kqueue - not found
[00:00:51] -- Looking for epoll_create
[00:00:53] -- Looking for epoll_create - not found
[00:00:53] -- Looking for sys/types.h
[00:00:55] -- Looking for sys/types.h - found
[00:00:55] -- Looking for stdint.h
[00:00:58] -- Looking for stdint.h - found
[00:00:58] -- Looking for stddef.h
[00:01:00] -- Looking for stddef.h - found
[00:01:00] -- Check size of struct pollfd
[00:01:02] -- Check size of struct pollfd - failed
[00:01:02] -- Looking for pollset_create
[00:01:04] -- Looking for pollset_create - not found
[00:01:04] -- Looking for poll
[00:01:06] -- Looking for poll - not found
[00:01:06] -- Using polling method in I/O threads: select
[00:01:06] -- Using polling method in zmq_poll(er)_* API: poll
[00:01:06] -- Detected ZMQ Version - 4.3.1
[00:01:06] -- Looking for include file ifaddrs.h
[00:01:09] -- Looking for include file ifaddrs.h - not found
[00:01:09] -- Looking for include file windows.h
[00:01:11] -- Looking for include file windows.h - found
[00:01:11] -- Looking for include file sys/uio.h
[00:01:13] -- Looking for include file sys/uio.h - not found
[00:01:13] -- Looking for include file sys/eventfd.h
[00:01:15] -- Looking for include file sys/eventfd.h - not found
[00:01:15] -- Looking for WSAStartup
[00:01:18] -- Looking for WSAStartup - found
[00:01:18] -- Looking for UuidCreateSequential
[00:01:20] -- Looking for UuidCreateSequential - found
[00:01:20] -- Looking for GetAdaptersAddresses
[00:01:23] -- Looking for GetAdaptersAddresses - found
[00:01:23] -- Looking for fopen in ws2
[00:01:25] -- Looking for fopen in ws2 - not found
[00:01:26] -- Looking for pthread.h
[00:01:28] -- Looking for pthread.h - not found
[00:01:28] -- Found Threads: TRUE  
[00:01:28] -- Looking for clock_gettime
[00:01:30] -- Looking for clock_gettime - not found
[00:01:30] -- Looking for fork
[00:01:32] -- Looking for fork - not found
[00:01:32] -- Looking for gethrtime
[00:01:34] -- Looking for gethrtime - not found
[00:01:34] -- Looking for mkdtemp
[00:01:36] -- Looking for mkdtemp - not found
[00:01:36] -- Looking for accept4
[00:01:38] -- Looking for accept4 - not found
[00:01:38] -- Performing Test HAVE_FLAG_/W3
[00:01:40] -- Performing Test HAVE_FLAG_/W3 - Success
[00:01:40] -- Performing Test HAVE_FLAG_-Wno-long-long
[00:01:42] -- Performing Test HAVE_FLAG_-Wno-long-long - Failed
[00:01:42] -- Performing Test HAVE_FLAG_-Wno-uninitialized
[00:01:44] -- Performing Test HAVE_FLAG_-Wno-uninitialized - Failed
[00:01:44] -- Performing Test HAVE_FLAG_-pedantic
[00:01:46] -- Performing Test HAVE_FLAG_-pedantic - Failed
[00:01:46] -- Performing Test HAVE_FLAG_-Werror
[00:01:48] -- Performing Test HAVE_FLAG_-Werror - Failed
[00:01:48] -- Performing Test HAVE_FLAG_/WX
[00:01:50] -- Performing Test HAVE_FLAG_/WX - Success
[00:01:50] CMake Warning (dev) at CMakeLists.txt:339 (if):
[00:01:50]   Policy CMP0054 is not set: Only interpret if() arguments as variables or
[00:01:50]   keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy
[00:01:50]   details.  Use the cmake_policy command to set the policy and suppress this
[00:01:50]   warning.
[00:01:50] 
[00:01:50]   Quoted variables like "MSVC" will no longer be dereferenced when the policy
[00:01:50]   is set to NEW.  Since the policy is not set the OLD behavior will be used.
[00:01:50] This warning is for project developers.  Use -Wno-dev to suppress it.
[00:01:50] 
[00:01:50] -- Performing Test HAVE_FLAG_-errwarn=%all
[00:01:52] -- Performing Test HAVE_FLAG_-errwarn=%all - Failed
[00:01:52] -- Checking whether SOCK_CLOEXEC is supported
[00:01:52] -- Performing Test ZMQ_HAVE_SOCK_CLOEXEC
[00:01:54] -- Performing Test ZMQ_HAVE_SOCK_CLOEXEC - Failed
[00:01:54] -- Checking whether O_CLOEXEC is supported
[00:01:54] -- Performing Test ZMQ_HAVE_O_CLOEXEC
[00:01:56] -- Performing Test ZMQ_HAVE_O_CLOEXEC - Failed
[00:01:56] -- Checking whether SO_BINDTODEVICE is supported
[00:01:56] -- Performing Test ZMQ_HAVE_SO_BINDTODEVICE
[00:01:58] -- Performing Test ZMQ_HAVE_SO_BINDTODEVICE - Failed
[00:01:58] -- Checking whether SO_KEEPALIVE is supported
[00:01:58] -- Performing Test ZMQ_HAVE_SO_KEEPALIVE
[00:02:00] -- Performing Test ZMQ_HAVE_SO_KEEPALIVE - Failed
[00:02:00] -- Checking whether TCP_KEEPCNT is supported
[00:02:00] -- Performing Test ZMQ_HAVE_TCP_KEEPCNT
[00:02:02] -- Performing Test ZMQ_HAVE_TCP_KEEPCNT - Failed
[00:02:02] -- Checking whether TCP_KEEPIDLE is supported
[00:02:02] -- Performing Test ZMQ_HAVE_TCP_KEEPIDLE
[00:02:04] -- Performing Test ZMQ_HAVE_TCP_KEEPIDLE - Failed
[00:02:04] -- Checking whether TCP_KEEPINTVL is supported
[00:02:04] -- Performing Test ZMQ_HAVE_TCP_KEEPINTVL
[00:02:06] -- Performing Test ZMQ_HAVE_TCP_KEEPINTVL - Failed
[00:02:06] -- Checking whether TCP_KEEPALIVE is supported
[00:02:06] -- Performing Test ZMQ_HAVE_TCP_KEEPALIVE
[00:02:08] -- Performing Test ZMQ_HAVE_TCP_KEEPALIVE - Failed
[00:02:08] -- Checking whether TIPC is supported
[00:02:08] -- Performing Test ZMQ_HAVE_TIPC
[00:02:10] -- Performing Test ZMQ_HAVE_TIPC - Failed
[00:02:10] -- Checking pthread_setname signature
[00:02:10] -- Performing Test ZMQ_HAVE_PTHREAD_SETNAME_1
[00:02:12] -- Performing Test ZMQ_HAVE_PTHREAD_SETNAME_1 - Failed
[00:02:12] -- Performing Test ZMQ_HAVE_PTHREAD_SETNAME_2
[00:02:14] -- Performing Test ZMQ_HAVE_PTHREAD_SETNAME_2 - Failed
[00:02:14] -- Performing Test ZMQ_HAVE_PTHREAD_SETNAME_3
[00:02:16] -- Performing Test ZMQ_HAVE_PTHREAD_SETNAME_3 - Failed
[00:02:16] -- Performing Test ZMQ_HAVE_PTHREAD_SET_NAME
[00:02:18] -- Performing Test ZMQ_HAVE_PTHREAD_SET_NAME - Failed
[00:02:18] -- Checking pthread_setaffinity signature
[00:02:18] -- Performing Test ZMQ_HAVE_PTHREAD_SETAFFINITY
[00:02:20] -- Performing Test ZMQ_HAVE_PTHREAD_SETAFFINITY - Failed
[00:02:20] -- Checking whether getrandom is supported
[00:02:20] -- Performing Test ZMQ_HAVE_GETRANDOM
[00:02:22] -- Performing Test ZMQ_HAVE_GETRANDOM - Failed
[00:02:22] -- Found PythonInterp: C:/Python27/python.exe (found version "2.7.15") 
[00:02:22] -- Could NOT find AsciiDoc (missing: ASCIIDOC_EXECUTABLE) 
[00:02:22] CMake Warning (dev) at tests/CMakeLists.txt:240 (message):
[00:02:22]   Test 'test_security_gssapi' is not known to CTest.
[00:02:22] This warning is for project developers.  Use -Wno-dev to suppress it.
[00:02:22] 
[00:02:22] -- Configuring done
[00:02:23] -- Generating done
[00:02:23] -- Build files have been written to: C:/projects/build_libzmq
```
This is much much slower than on Linux, but I don't know how to speed this up. I suffer from this locally as well. It seems to be a general cmake issue with MSVC++. Maybe we can find some help with the cmake guys.

Reply:
One option is to skip some of the tests, which cannot possibly succeed with MSVC (there is never pthread etc.). I can add this to the CMakeLists

Reply:
What about enabled "Rolling builds" on appveyor, so that builds in a PR are automatically cancelled when there is a new push?

Reply:
Yes it is quite buried. I knew that there was such an option, but looked through the list of options twice until I found it... and there is no explanation either.

Reply:
cmake run is down to under a minute now (still much slower than under Linux :( ):
```
[00:00:12] cmake -D CMAKE_INCLUDE_PATH="%SODIUM_INCLUDE_DIR%" -D CMAKE_LIBRARY_PATH="%SODIUM_LIBRARY_DIR%" -D WITH_LIBSODIUM="%WITH_LIBSODIUM%" -D ENABLE_DRAFTS="ON" -D ENABLE_CURVE="%ENABLE_CURVE%" -D API_POLLER="%API_POLLER%" -D POLLER="%POLLER%" -D CMAKE_C_FLAGS_RELEASE="/MT" -D CMAKE_C_FLAGS_DEBUG="/MTd" -D WITH_LIBSODIUM="%WITH_LIBSODIUM%" -D LIBZMQ_WERROR="ON" -G "%CMAKE_GENERATOR%" "%APPVEYOR_BUILD_FOLDER%"
[00:00:17] -- The C compiler identification is MSVC 18.0.40629.0
[00:00:19] -- The CXX compiler identification is MSVC 18.0.40629.0
[00:00:19] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe
[00:00:22] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe -- works
[00:00:22] -- Detecting C compiler ABI info
[00:00:25] -- Detecting C compiler ABI info - done
[00:00:25] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe
[00:00:27] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe -- works
[00:00:27] -- Detecting CXX compiler ABI info
[00:00:29] -- Detecting CXX compiler ABI info - done
[00:00:29] -- Detecting CXX compile features
[00:00:32] -- Detecting CXX compile features - done
[00:00:32] -- Performing Test COMPILER_SUPPORTS_CXX11
[00:00:34] -- Performing Test COMPILER_SUPPORTS_CXX11 - Failed
[00:00:34] -- Performing Test COMPILER_SUPPORTS_C11
[00:00:37] -- Performing Test COMPILER_SUPPORTS_C11 - Failed
[00:00:37] -- Found SODIUM: C:/projects/libsodium/bin/Win32/Release/v120/dynamic/libsodium.lib  
[00:00:37] -- Using libsodium for CURVE security
[00:00:37] -- Using polling method in I/O threads: select
[00:00:37] -- Using polling method in zmq_poll(er)_* API: select
[00:00:37] -- Detected ZMQ Version - 4.3.1
[00:00:37] -- Looking for include file ifaddrs.h
[00:00:39] -- Looking for include file ifaddrs.h - not found
[00:00:39] -- Looking for include file windows.h
[00:00:42] -- Looking for include file windows.h - found
[00:00:42] -- Looking for include file sys/uio.h
[00:00:44] -- Looking for include file sys/uio.h - not found
[00:00:44] -- Looking for include file sys/eventfd.h
[00:00:46] -- Looking for include file sys/eventfd.h - not found
[00:00:46] -- Looking for WSAStartup
[00:00:49] -- Looking for WSAStartup - found
[00:00:49] -- Looking for UuidCreateSequential
[00:00:51] -- Looking for UuidCreateSequential - found
[00:00:51] -- Looking for GetAdaptersAddresses
[00:00:54] -- Looking for GetAdaptersAddresses - found
[00:00:54] -- Looking for fopen in ws2
[00:00:56] -- Looking for fopen in ws2 - not found
[00:00:57] -- Looking for pthread.h
[00:00:59] -- Looking for pthread.h - not found
[00:00:59] -- Found Threads: TRUE  
[00:00:59] -- Performing Test HAVE_FLAG_/W3
[00:01:01] -- Performing Test HAVE_FLAG_/W3 - Success
[00:01:01] -- Performing Test HAVE_FLAG_/WX
[00:01:04] -- Performing Test HAVE_FLAG_/WX - Success
[00:01:04] -- Found PythonInterp: C:/Python27/python.exe (found version "2.7.15") 
[00:01:04] -- Could NOT find AsciiDoc (missing: ASCIIDOC_EXECUTABLE) 
[00:01:04] CMake Warning (dev) at tests/CMakeLists.txt:243 (message):
[00:01:04]   Test 'test_security_gssapi' is not known to CTest.
[00:01:04] This warning is for project developers.  Use -Wno-dev to suppress it.
[00:01:04] 
[00:01:04] -- Configuring done
[00:01:05] -- Generating done
[00:01:05] -- Build files have been written to: C:/projects/build_libzmq
```